compilation/types/source_maps.go

```diff
@@ -0,0 +1,174 @@
+// https://docs.soliditylang.org/en/latest/internals/source_mappings.html
+
+// SourceMapJumpType describes the type of jump operation occurring within a SourceMapElement if the instruction
+// is jumping.
+type SourceMapJumpType string
+
+const (
+	// SourceMapJumpTypeNone indicates no jump occurred.
+	SourceMapJumpTypeNone SourceMapJumpType = ""
+
+	// SourceMapJumpTypeJumpIn indicates a jump into a function occurred.
+	SourceMapJumpTypeJumpIn SourceMapJumpType = "i"
+
+	// SourceMapJumpTypeJumpOut indicates a return from a function occurred.
+	SourceMapJumpTypeJumpOut SourceMapJumpType = "o"
+
+	// SourceMapJumpTypeJumpWithin indicates a jump occurred within the same function, e.g. for loops.
+	SourceMapJumpTypeJumpWithin SourceMapJumpType = "-"
+)
+
+// SourceMap describes a list of elements which correspond to instruction indexes in compiled bytecode, describing
+// which source files and the start/end range of the source code which the instruction maps to.
+type SourceMap []SourceMapElement
+
+// SourceMapElement describes an individual element of a source mapping output by the compiler.
+// The index of each element in a source map corresponds to an instruction index (not to be mistaken with offset).
+// It describes portion of a source file the instruction references.
+type SourceMapElement struct {
+	// Offset refers to the byte offset which marks the start of the source range the instruction maps to.
+	Offset int
+
+	// Length refers to the byte length of the source range the instruction maps to.
+	Length int
+
+	// FileID refers to an identifier for the CompiledSource file which houses the relevant source code.
+	FileID int
+
+	// JumpType refers to the SourceMapJumpType which provides information about any type of jump that occurred.
+	JumpType SourceMapJumpType
+
+	// ModifierDepth refers to the depth in which code has executed a modifier function. This is used to assist
+	// debuggers, e.g. understanding if the same modifier is re-used multiple times in a call.
+	ModifierDepth int
+}
+
+// ParseSourceMap takes a source mapping string returned by the compiler and parses it into an array of
+// SourceMapElement objects.
+// Returns the list of SourceMapElement objects.
+func ParseSourceMap(sourceMapStr string) (SourceMap, error) {
+	// Define our variables to store our results in
+	var (
+		sourceMap SourceMap
+		err       error
+	)
+
+	// If our provided source map string is empty, there is no work to be done.
+	if len(sourceMapStr) == 0 {
+		return sourceMap, nil
+	}
+
+	// Separate all the individual source mapping elements
+	elements := strings.Split(sourceMapStr, ";")
+
+	// We use this variable to store "the previous element" because the way
+	// the source mapping works when an element or field is "empty"
+	// the value of the previous element is used.
+	previous := SourceMapElement{
+		Offset:        -1,
+		Length:        -1,
+		FileID:        -1,
```

Comment:
This would largely be a no-op for medusa, with the exception that if a non-empty source map string is supplied without enough "parts" for any element, we might want to throw an error (not allow a valid offset but length of -1). Otherwise, FileID of -1 will be a no-op for how medusa uses it, but for an API consideration, we should also probably throw an error if it's not ever supplied.

I'll resolve this later.